### PR TITLE
Fix issue with convert function when converting string to date, datetimeoffset, datetime2, datetime, smalldatetime or time

### PR DESCRIPTION
**Pull request by rohit01010**
### Description
This PR will add fix for following issues:
- For converting string to datetime2, datetimeoffset and smalldatetime using convert function, we are directly using cast function which will convert the string to given type with default style. Added support to use corresponding SQL helper functions defined which will fix this issue.
- For conversion of string to date, datetime, smalldatetime and time using convert function, there are helper functions defined but are unable to handle all the supported datetime string literals. Refactored these definitions and fixed this issue.
- For conversion of string to datetime2 and datetimeoffset using convert function, there are no helper functions defined which handles this conversion as there are for conversion of string to date, datetime, smalldatetime and time. Added definitions to handle these datatypes as well.
- Function sys.babelfish_get_microsecs_from_fractsecs() is unable to handle the edge case where fractional seconds get overflowed after rounding it to given scale.
- Function babelfish_get_timeunit_from_string only returns HOURS, MINUTES, SECONDS and FRACTSECONDS from a given time string. Extended this function to return OFFHOURS, OFFMINUTES and OFFSIGN to get details of timeoffset from time string.

First commit changes cherry-picked from: https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2613

Authored-by: Rohit Bhagat [rohitbgt@amazon.com](mailto:rohitbgt@amazon.com)
Signed-off-by: Rohit Bhagat [rohitbgt@amazon.com](mailto:rohitbgt@amazon.com)

### Issues Resolved
BABEL-4896

### Test Scenarios Covered ###
* **Use case based -** YES

* **Boundary conditions -** YES

* **Arbitrary inputs -** YES

* **Negative test cases -** YES

* **Minor version upgrade tests -** YES

* **Major version upgrade tests -** YES

* **Performance tests -** NA

* **Tooling impact -** NA

* **Client tests -** YES

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/sql/sys_function_helpers.sql
Comment (rishabhtanwar29) [inline contrib/babelfishpg_tsql/sql/sys_function_helpers.sql]: Nit: `Following condition will handle overflow of fractsecs`

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c
Comment (rishabhtanwar29) [inline contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:206]: Can be re-written as: `apply typmod cast on top of CONVERT`

File: test/JDBC/upgrade/16_2/schedule
Comment (rishabhtanwar29) [inline test/JDBC/upgrade/16_2/schedule]: Why have you only added these upto 16.2, 15.6 and 14.11? Why not add these all previous versions?

File: test/JDBC/upgrade/15_7/schedule
Comment (rishabhtanwar29) [inline test/JDBC/upgrade/15_7/schedule]: We are going to backpatch this to 15 and 14 as well so then ideally we shouldn't be using `-before` suffix in 15_7 and 14_12 schedule files but it will fail until we actually merge this change to those versions. So, let's make sure to have PRs for the previous versions merge them from 14 to 16 order.

File: contrib/babelfishpg_tsql/sql/sys_function_helpers.sql
Comment (rohit01010) [reply contrib/babelfishpg_tsql/sql/sys_function_helpers.sql]: Done

File: test/JDBC/upgrade/16_2/schedule
Comment (rohit01010) [reply test/JDBC/upgrade/16_2/schedule]: Added in all the base versions

File: test/JDBC/upgrade/15_7/schedule
Comment (rohit01010) [reply test/JDBC/upgrade/15_7/schedule]: yes, we will merge this PR, once the PR's from previous version's gets merged

**Review by rishabhtanwar29 — APPROVED** [review]

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:206]: existing comment looks simple and easy to understand, let's not change it.

**Review by shardgupta — APPROVED** [review]

**Pull request merged**
